sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1309,3 +1311,20 @@ object RemoveRepetitionFromGroupExpressions extends Rule[LogicalPlan] {
       }
   }
 }
+
+/**
+ * Updates nullability in [[AttributeReference]]s if nullability is different between
+ * non-leaf plan's expressions and the children output.
+ */
+object UpdateNullabilityInAttributeReferences extends Rule[LogicalPlan] {
```

Comment:
this rule itself is useful without the filter optimization.

In general, I don't think it's useful to optimize filter using the `IsNotNull` expression, as it's not a common case. I think we can update `Join.output` to specify nullability for join keys, which seems easier and a more common case.

Reply:
ok, I'll make this pr simpler (I'll drop some changes for filters).

Reply:
I dropped the changes of `execution.FilterExec` though, you suggested we would drop the changes of `logical.Filter`, too? https://github.com/apache/spark/pull/18576/files#diff-72917e7b68f0311b2fb42990e0dc616dR139

I basically agree that the `Join.output` modification is more simple/important, but is it okay to ignore nullability in `logical.Filter`? For example, in the current master, `QueryPlanConstraints.inferIsNotNullConstraints` appends non-nullable constraints in `logical.Filter` and this constraints aren't correctly propagated into upper plan nodes now. So, I think it'd be better to respect nullability in both `logical.Join` and `logical.Filter`.

Reply:
If we think more about how these `IsNotNull` constraints are generated, they mostly come from join. I don't think it's a common case users put `IsNotNull` in filter.

Reply:
ok, I'll drop it in this pr. just a sec.